File: test-e2e.js
// End-to-end test: start backend, instrument functions, query with CLI
const { trickle, configure, flush } = require('./packages/client-js/dist/index');

configure({ backendUrl: 'http://localhost:4888', batchIntervalMs: 500, debug: true });

// Simulate an Express-style handler
const processOrder = trickle(function processOrder(order) {
  const total = order.items.reduce((sum, item) => sum + item.price * item.quantity, 0);
  const tax = total * 0.1;
  return {
    orderId: order.id,
    customer: order.customer.name,
    total,
    tax,
    grandTotal: total + tax,
    status: 'processed',
  };
});

// Simulate a function that errors
const validatePayment = trickle(function validatePayment(payment) {
  if (!payment.cardNumber) {
    throw new Error('Card number is required');
  }
  if (payment.amount <= 0) {
    throw new TypeError('Amount must be positive');
  }
  return { valid: true, method: payment.method };
});

// Simulate a user lookup
const getUser = trickle(function getUser(userId, options) {
  return {
    id: userId,
    name: 'John Doe',
    email: 'john@example.com',
    roles: ['admin', 'user'],
    metadata: {
      lastLogin: new Date().toISOString(),
      loginCount: 42,
    },
  };
});

async function run() {
  console.log('=== Running instrumented functions ===\n');

  // Happy path calls
  console.log('1. processOrder (happy path)');
  const result = processOrder({
    id: 'ORD-123',
    customer: { name: 'Alice', email: 'alice@example.com' },
    items: [
      { name: 'Widget', price: 29.99, quantity: 2 },
      { name: 'Gadget', price: 49.99, quantity: 1 },
    ],
  });
  console.log('   Result:', JSON.stringify(result));

  console.log('2. getUser (happy path)');
  const user = getUser('usr-456', { includeRoles: true });
  console.log('   Result:', JSON.stringify(user));

  // Sad path: trigger errors
  console.log('3. validatePayment (error: missing card)');
  try {
    validatePayment({ amount: 100, method: 'credit' });
  } catch (e) {
    console.log('   Caught:', e.message);
  }

  console.log('4. validatePayment (error: bad amount)');
  try {
    validatePayment({ cardNumber: '4111111111111111', amount: -5, method: 'credit' });
  } catch (e) {
    console.log('   Caught:', e.message);
  }

  console.log('5. validatePayment (happy path)');
  const valid = validatePayment({ cardNumber: '4111111111111111', amount: 50, method: 'debit' });
  console.log('   Result:', JSON.stringify(valid));

  // Explicitly flush to ensure all data reaches the backend
  console.log('\nFlushing data to backend...');
  await flush();
  console.log('Done! Now query with the CLI.\n');
}

run().catch(console.error);

File: test-e2e.py
"""End-to-end test for the Python trickle client."""
import sys
sys.path.insert(0, 'packages/client-python/src')

from trickle import trickle, configure

configure(backend_url='http://localhost:4888', batch_interval=0.5)


@trickle
def process_order(order):
    total = sum(item['price'] * item['quantity'] for item in order['items'])
    tax = total * 0.1
    return {
        'order_id': order['id'],
        'customer': order['customer']['name'],
        'total': total,
        'tax': tax,
        'grand_total': total + tax,
        'status': 'processed',
    }


@trickle
def validate_payment(payment):
    if not payment.get('card_number'):
        raise ValueError('Card number is required')
    if payment['amount'] <= 0:
        raise TypeError('Amount must be positive')
    return {'valid': True, 'method': payment['method']}


@trickle
def get_user(user_id, include_roles=False):
    return {
        'id': user_id,
        'name': 'Jane Smith',
        'email': 'jane@example.com',
        'roles': ['admin', 'editor'] if include_roles else [],
        'metadata': {
            'last_login': '2026-03-10T10:00:00Z',
            'login_count': 99,
        },
    }


def main():
    import time

    print('=== Running Python instrumented functions ===\n')

    print('1. process_order (happy path)')
    result = process_order({
        'id': 'ORD-789',
        'customer': {'name': 'Bob', 'email': 'bob@example.com'},
        'items': [
            {'name': 'Book', 'price': 15.99, 'quantity': 3},
            {'name': 'Pen', 'price': 2.50, 'quantity': 10},
        ],
    })
    print(f'   Result: {result}')

    print('2. get_user (happy path)')
    user = get_user('usr-101', include_roles=True)
    print(f'   Result: {user}')

    print('3. validate_payment (error: missing card)')
    try:
        validate_payment({'amount': 100, 'method': 'credit'})
    except ValueError as e:
        print(f'   Caught: {e}')

    print('4. validate_payment (error: bad amount)')
    try:
        validate_payment({'card_number': '5500000000000004', 'amount': -10, 'method': 'debit'})
    except TypeError as e:
        print(f'   Caught: {e}')

    print('5. validate_payment (happy path)')
    result = validate_payment({'card_number': '5500000000000004', 'amount': 75, 'method': 'debit'})
    print(f'   Result: {result}')

    print('\nWaiting for flush...')
    time.sleep(3)
    print('Done!')


if __name__ == '__main__':
    main()
